File: frontend/src/types/dashboards.ts
import { z } from "zod";

export const WidgetTypeSchema = z.enum(["traffic_flow"]);
export type WidgetType = z.infer<typeof WidgetTypeSchema>;

export const DashboardWidgetSchema = z.object({
  id: z.string(),
  type: WidgetTypeSchema,
  config: z.record(z.string(), z.unknown()).default({}),
});
export type DashboardWidget = z.infer<typeof DashboardWidgetSchema>;

export const DashboardSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string().nullable(),
  widgets: z.array(DashboardWidgetSchema),
  created_at: z.string(),
  updated_at: z.string(),
});
export type Dashboard = z.infer<typeof DashboardSchema>;

export type DashboardCreatePayload = {
  name: string;
  description?: string | null;
};

export type DashboardUpdatePayload = {
  name?: string;
  description?: string | null;
  widgets?: DashboardWidget[];
};

export type DashboardWidgetCreatePayload = {
  type: WidgetType;
  config?: Record<string, unknown>;
};

File: backend/app/schemas/dashboards.py
from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class WidgetType(StrEnum):
    TRAFFIC_FLOW = "traffic_flow"


class DashboardWidget(BaseModel):
    id: str = Field(min_length=1, max_length=36)
    type: WidgetType
    config: dict[str, Any] = Field(default_factory=dict)


class DashboardCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=2000)


class DashboardUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=2000)
    widgets: list[DashboardWidget] | None = None


class DashboardWidgetCreate(BaseModel):
    type: WidgetType
    config: dict[str, Any] = Field(default_factory=dict)


class DashboardRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    description: str | None
    widgets: list[DashboardWidget]
    created_at: datetime
    updated_at: datetime
